code/modules/lighting/lighting_source.dm

```diff
@@ -0,0 +1,294 @@
+/datum/light_source
+	var/atom/top_atom       // The atom we're emitting light from (for example a mob if we're from a flashlight that's being held).
+	var/atom/source_atom    // The atom that we belong to.
+
+	var/turf/source_turf    // The turf under the above.
+	var/light_power         // Intensity of the emitter light.
+	var/light_range         // The range of the emitted light.
+	var/light_color         // The colour of the light, string, decomposed by parse_light_color()
+
+	// Variables for keeping track of the colour.
+	var/lum_r
+	var/lum_g
+	var/lum_b
+
+	// The lumcount values used to apply the light.
+	var/tmp/applied_lum_r
+	var/tmp/applied_lum_g
+	var/tmp/applied_lum_b
+
+	var/list/datum/lighting_corner/effect_str     // List used to store how much we're affecting corners.
+	var/list/turf/affecting_turfs
+
+	var/applied             // Whether we have applied our light yet or not.
+
+	var/vis_update          // Whether we should smartly recalculate visibility. and then only update tiles that became (in)visible to us.
+	var/needs_update        // Whether we are queued for an update.
+	var/destroyed           // Whether we are destroyed and need to stop emitting light.
+	var/force_update
+
+/datum/light_source/New(var/atom/owner, var/atom/top)
+	source_atom = owner // Set our new owner.
+	if (!source_atom.light_sources)
+		source_atom.light_sources = list()
+
+	source_atom.light_sources += src // Add us to the lights of our owner.
+	top_atom = top
+	if (top_atom != source_atom)
+		if (!top.light_sources)
+			top.light_sources     = list()
+
+		top_atom.light_sources += src
+
+	source_turf = top_atom
+	light_power = source_atom.light_power
+	light_range = source_atom.light_range
+	light_color = source_atom.light_color
+
+	parse_light_color()
+
+	effect_str      = list()
+	affecting_turfs = list()
+
+	update()
+
+	return ..()
+
+// Kill ourselves.
+/datum/light_source/proc/destroy()
+	destroyed = TRUE
+	force_update()
+	if (source_atom)
+		source_atom.light_sources -= src
+
+	if (top_atom)
+		top_atom.light_sources    -= src
+
+#ifdef LIGHTING_INSTANT_UPDATES
+/datum/light_source/proc/effect_update()
+	if (check() || destroyed || force_update)
+		remove_lum()
+		if (!destroyed)
+			apply_lum()
+
+	else if (vis_update)	// We smartly update only tiles that became (in) visible to use.
+		smart_vis_update()
+
+	vis_update   = FALSE
+	force_update = FALSE
+	needs_update = FALSE
+#else
+
+// Call it dirty, I don't care.
+// This is here so there's no performance loss on non-instant updates from the fact that the engine can also do instant updates.
+// If you're wondering what's with the "BYOND" argument: BYOND won't let me have a () macro that has no arguments :|.
```

Comment:
@PJB3005 why does this need to be a () macro at all
wouldn't it work if it was just a regular insertion macro?